### PR TITLE
nixos/restic: add option to inhibit going to sleep

### PR DESCRIPTION
**Pull request by cheriimoya**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/doc/manual/release-notes/rl-2405.section.md [hunk 1]
```diff
@@ -394,6 +394,8 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
 
   `redirectCode`.
 
+- `restic` module now has an option for inhibiting system sleep while backups are running, defaulting to off (not inhibiting sleep), available as [`services.restic.backups.<name>.inhibitsSleep`](#opt-services.restic.backups._name_.inhibitsSleep).
```
Comment (lolbinarycat) [inline nixos/doc/manual/release-notes/rl-2405.section.md:397]: this documents the option more clearly than the option's actual documentation, although `false = don't inhibit` should probably be obvious from the option name, so i its probably fine as-is.

File: nixos/modules/services/backup/restic.nix [hunk 2]
```diff
@@ -292,7 +301,14 @@ in
...
           let
             extraOptions = concatMapStrings (arg: " -o ${arg}") backup.extraOptions;
-            resticCmd = "${backup.package}/bin/restic${extraOptions}";
+            inhibitCmd = concatStringsSep " " [
+              "${pkgs.systemd}/bin/systemd-inhibit"
+              "--mode='block'"
+              "--who='restic'"
+              "--what='sleep'"
+              "--why='Scheduled backup ${name}' "
+            ];
```
Comment (MinerSebas) [inline nixos/modules/services/backup/restic.nix:310]: Why split the string beforhand?
```suggestion
            inhibitCmd = "${pkgs.systemd}/bin/systemd-inhibit --mode='block' --who='restic' --what='sleep' --why='Scheduled backup ${name}' "
            ];
```

Comment (cheriimoya) [reply nixos/modules/services/backup/restic.nix:310]: I split it up for readability, no other reason. I could change it, but the next reviewer would probably tell me to split it up as it is too long:D

File: nixos/doc/manual/release-notes/rl-2405.section.md (hunk 1 above)
Comment (cheriimoya) [reply nixos/doc/manual/release-notes/rl-2405.section.md:397]: What do you think about the updated option description?

File: nixos/modules/services/backup/restic.nix [hunk 3]
```diff
@@ -292,7 +304,14 @@ in
         (name: backup:
           let
             extraOptions = concatMapStrings (arg: " -o ${arg}") backup.extraOptions;
-            resticCmd = "${backup.package}/bin/restic${extraOptions}";
+            inhibitCmd = concatStringsSep " " [
+              "${pkgs.systemd}/bin/systemd-inhibit"
+              "--mode='block'"
+              "--who='restic'"
+              "--what='sleep'"
+              "--why='Scheduled backup ${name}' "
```
Comment (networkException) [inline nixos/modules/services/backup/restic.nix:312]:
```suggestion
              "--why=${escapeShellArg "Scheduled backup ${name}"} "
```

The module isn't great at escaping `name` but we should make sure that it doesn't spill over from `systemd-inhibit`s args parsing here.

**Review by networkException — APPROVED** [review]
Looks good to me, needs another rebase though

**Comment by emilylange** [comment]
@ofborg test restic

**Comment by emilylange** [comment]
The newly added VM test section for this seems to fail, but the VM test itself reports as successful anyway.
Can you look into this?

```
server # [   68.052903] systemd[1]: Starting restic-backups-inhibit-test.service...
server: waiting for success: systemd-inhibit --no-legend --no-pager | grep -q restic
(finished: waiting for success: systemd-inhibit --no-legend --no-pager | grep -q restic, in 0.08 seconds)
(finished: run the VM test script, in 68.89 seconds)
server # [   68.177458] restic-backups-inhibit-test-pre-start[1788]: Fatal: unable to open config file: stat /root/restic-backup-inhibit-test/config: no such file or directory
server # [   68.180268] restic-backups-inhibit-test-pre-start[1788]: Is there a repository at the following location?
server # [   68.181806] restic-backups-inhibit-test-pre-start[1788]: /root/restic-backup-inhibit-test
server # [   68.186114] restic-backups-inhibit-test-pre-start[1785]: /nix/store/3jwc6w0dbsfwcnl4d91ssq77x8phyihd-restic-0.16.5/bin/restic failed with exit status 1.
test script finished in 68.96s
```

See https://logs.ofborg.org/?key=nixos/nixpkgs.296691&attempt_id=d81aedde-c33a-40c0-9060-b638f8a1048a

Also, what is the benefit of wrapping the command with `systemd-inhitbit` instead of preventing sleep like this:

```diff
diff --git a/nixos/modules/services/backup/restic.nix b/nixos/modules/services/backup/restic.nix
index 7b10d7d38d20..903eb9ceeafc 100644
--- a/nixos/modules/services/backup/restic.nix
+++ b/nixos/modules/services/backup/restic.nix
@@ -308,14 +308,7 @@ in
         (name: backup:
           let
             extraOptions = concatMapStrings (arg: " -o ${arg}") backup.extraOptions;
-            inhibitCmd = concatStringsSep " " [
-              "${pkgs.systemd}/bin/systemd-inhibit"
-              "--mode='block'"
-              "--who='restic'"
-              "--what='sleep'"
-              "--why=${escapeShellArg "Scheduled backup ${name}"} "
-            ];
-            resticCmd = "${optionalString backup.inhibitsSleep inhibitCmd}${backup.package}/bin/restic${extraOptions}";
+            resticCmd = "${backup.package}/bin/restic${extraOptions}";
             excludeFlags = optional (backup.exclude != []) "--exclude-file=${pkgs.writeText "exclude-patterns" (concatStringsSep "\n" backup.exclude)}";
             filesFromTmpFile = "/run/restic-backups-${name}/includes";
             doBackup = (backup.dynamicFilesFrom != null) || (backup.paths != null && backup.paths != []);
@@ -353,6 +346,18 @@ in
             restartIfChanged = false;
             wants = [ "network-online.target" ];
             after = [ "network-online.target" ];
+            before = optionals backup.inhibitsSleep [
+              "suspend.target"
+              "hibernate.target"
+              "hybrid-sleep.target"
+              "suspend-then-hibernate.target"
+            ];
+            wantedBy = optionals backup.inhibitsSleep [
+              "suspend.target"
+              "hibernate.target"
+              "hybrid-sleep.target"
+              "suspend-then-hibernate.target"
+            ];
             serviceConfig = {
               Type = "oneshot";
               ExecStart = (optionals doBackup [ "${resticCmd} backup ${concatStringsSep " " (backup.extraBackupArgs ++ excludeFlags)} --files-from=${filesFromTmpFile}" ])
```

I would assume it does not show up in `systemd-inhibit` that way, but is that really of that much use?

**Comment by emilylange** [comment]
> What is an advantage of the `before`/`wantedBy` relation over the wrapped version?

The current wrapped version is silently failing the test, so I wanted to provide another approach that I know works.

> Another benefit is, that if I execute `systemctl suspend`, systemd tells me that there is an inhibitor instead of not giving any output and seemingly hanging.

Hmm I see. That's a solid reason. Keep your `systemd-inhibit` approach then. Sounds good.

Feel free to undraft and ping when you fixed the VM test :)

**Comment by cheriimoya** [comment]
Okay, after some debugging I found the cause of the exit 1 output...
It's the output of `systemd-inhibit`.
The `initialize` option does a `resticCmd snapshot || resticCmd init` and as the first command fails if the repo is not initialized, the systemd-inhibit version fails verbosely than the version without.
I could leave it as is or alter `resticCmd snapshot || resticCmd init` to `resticCmd snapshot ${lib.optionalString backup.inhibit "2> /dev/null"} || resticCmd init`.
What do you think?

```
$ RESTIC_REPOSITORY=/none/restic-backup-inhibit-test systemd-inhibit /nix/store/0lvsqlp1n5c0mckcglfbxih43x0amysi-restic-0.16.4/bin/restic snapshots                                                       [9:43:52]
Fatal: unable to open config file: stat /none/restic-backup-inhibit-test/config: no such file or directory
Is there a repository at the following location?
/none/restic-backup-inhibit-test
/nix/store/0lvsqlp1n5c0mckcglfbxih43x0amysi-restic-0.16.4/bin/restic failed with exit status 1.
[1]    2674739 exit 1     RESTIC_REPOSITORY=/none/restic-backup-inhibit-test systemd-inhibit  snapshots
$ RESTIC_REPOSITORY=/none/restic-backup-inhibit-test /nix/store/0lvsqlp1n5c0mckcglfbxih43x0amysi-restic-0.16.4/bin/restic snapshots                                                                       [9:43:54]
Fatal: unable to open config file: stat /none/restic-backup-inhibit-test/config: no such file or directory
Is there a repository at the following location?
/none/restic-backup-inhibit-test
[1]    2674760 exit 1     RESTIC_REPOSITORY=/none/restic-backup-inhibit-test  snapshots
```

image

File: nixos/tests/restic.nix [hunk 4]
```diff
@@ -190,6 +197,13 @@ import ./make-test-python.nix (
           'restic-remotebackup snapshots --json | ${pkgs.jq}/bin/jq "length | . == 1"',
 
       )
+
+      # test that the inhibit option is working
+      server.systemctl("start --no-block restic-backups-inhibit-test.service")
+      server.wait_until_succeeds(
+          "systemd-inhibit --no-legend --no-pager | grep -q restic",
+          5
```
Comment (emilylange) [inline nixos/tests/restic.nix:205]: Wondering if a 5 second timeout is too low for hydra.nixos.org.

I suppose we will see how it goes and can create a follow-up PR if needed :)

**Review by emilylange — APPROVED** [review]
Ah I see, this is a misunderstanding then.

The error line comes from `resticCmd snapshot` (which is intended) and just happens to be the last line the VM test prints. 

Makes sense to me now!

**Comment by emilylange** [comment]
@ofborg test restic

**Pull request merged**